lib/mix/lib/mix/dep/lock.ex

```diff
@@ -32,48 +32,78 @@ defmodule Mix.Dep.Lock do
   """
   @spec read() :: map
   def read() do
-    case File.read(lockfile()) do
-      {:ok, info} ->
-        assert_no_merge_conflicts_in_lockfile(lockfile(), info)
+    lockfile = lockfile()
 
-        case Code.eval_string(info, [], file: lockfile()) do
-          {lock, _binding} when is_map(lock) -> lock
-          {_, _binding} -> %{}
-        end
+    case File.read(lockfile) do
+      {:ok, content} ->
+        read_lock(content, lockfile)
 
       {:error, _} ->
         %{}
     end
   end
 
+  @conflict_delimiters ["<<<<<<<", "=======", ">>>>>>>"]
+
+  defp read_lock(content, lockfile) do
+    read_lock(content, lockfile, String.contains?(content, @conflict_delimiters))
+  end
+
+  @conflict_regex ~r/<{7} .+\n|\|{7} .+\n[\w\W\s]+={7}\n|={7}\n|>{7} .+\n/
+
+  defp read_lock(content, lockfile, true) do
+    case content |> String.replace(@conflict_regex, "") |> eval_content(lockfile) do
+      {:ok, new_map} ->
+        write_lock(new_map, %{}, lockfile)
+        new_map
+
+      :error ->
+        Mix.raise("""
+        Your #{lockfile} contains merge conflicts we cannot automatically solve.
+        Please resolve the conflicts manually and run the command again
+        """)
+    end
+  end
+
+  defp read_lock(content, lockfile, false) do
+    case eval_content(content, lockfile) do
+      {:ok, map} -> map
+      :error -> Mix.raise("Could not read #{lockfile} due to a syntax error.")
+    end
+  end
+
+  defp eval_content(content, lockfile) do
+    case Code.eval_string(content, [], file: lockfile) do
+      {lock, _binding} when is_map(lock) -> {:ok, lock}
+      {_, _binding} -> {:ok, %{}}
+    end
+  rescue
+    SyntaxError -> :error
+  end
+
   @doc """
   Receives a map and writes it as the latest lock.
   """
   @spec write(map) :: :ok
   def write(map) do
-    unless map == read() do
```

Comment:
Please revert those changes too. They are touching an unrelated part of the PR.

Reply:
@josevalim, I only did this because I reused it on https://github.com/elixir-lang/elixir/pull/7094/files#diff-ba2dc6a9a3e9c51bdfd55a2b51510d4eR57. Should we duplicate this code to keep `Mix.Dep.Lock.write/1` untouched?

Reply:
I see. It makes sense now, please leave it as is then, thanks!